File: leetcode/242.valid-anagram/ts/isanagram.ts
// N, K
const isAnagram = (s: string, t: string): boolean => {
    if (t.length !== s.length) return false

    const characterCounts = new Map<string, number>()

    for (const c of s) {
        const v = characterCounts.get(c) ?? 0
        characterCounts.set(c, v + 1)
    }

    for (const c of t) {
        const v = characterCounts.get(c) ?? 0
        if (v <= 0) return false
        characterCounts.set(c, v - 1)
    }

    return true
};

export default isAnagram

File: leetcode/242.valid-anagram/cpp/isanagram.cpp
#include <string>
#include <unordered_map>
#include <array>

class Solution
{
public:
    // N, K
    bool isAnagram(std::string s, std::string t)
    {
        return isAnagramArray(s, t);
    }

private:
    bool isAnagramArray(std::string s, std::string t)
    {
        if (s.length() != t.length())
            return false;

        std::array<int, 26> arr{};

        for (auto c : s)
            arr[dist(c)]++;

        for (auto c : t)
        {
            auto i = dist(c);

            arr[i]--;
            if (arr[i] < 0)
                return false;
        }
        return true;
    };

    int dist(char ch)
    {
        return ch - 'a';
    }

    bool isAnagramBasic(std::string s, std::string t)
    {
        if (s.length() != t.length())
            return false;

        auto characterCounts = std::unordered_map<char, int>{26};

        for (auto c : s)
            characterCounts[c]++;

        for (auto c : t)
        {
            characterCounts[c]--;
            if (characterCounts[c] < 0)
                return false;
        }
        return true;
    }
};
